src/vs/editor/contrib/copyPaste/browser/copyPasteController.ts

```diff
@@ -0,0 +1,188 @@
+
+		const container = editor.getContainerDomNode();
+
+		this._register(addDisposableListener(container, 'copy', (e: ClipboardEvent) => {
+			if (!e.clipboardData) {
+				return;
+			}
+
+			const model = editor.getModel();
+			const selection = this._editor.getSelection();
+			if (!model || !selection) {
+				return;
+			}
+
+			if (!this.arePasteActionsEnabled(model)) {
+				return;
+			}
+
+			const providers = this._languageFeaturesService.documentPasteEditProvider.ordered(model).filter(x => !!x.prepareDocumentPaste);
+			if (!providers.length) {
+				return;
+			}
+
+			const dataTransfer = toVSDataTransfer(e.clipboardData);
+
+			// Save off a handle pointing to data that VS Code maintains.
+			const handle = generateUuid();
+			e.clipboardData.setData(vscodeClipboardMime, handle);
+
+			const promise = createCancelablePromise(async token => {
+				const results = await Promise.all(providers.map(provider => {
+					return provider.prepareDocumentPaste!(model, selection, dataTransfer, token);
+				}));
+
+				for (const result of results) {
+					result?.forEach((value, key) => {
+						dataTransfer.set(key, value);
+					});
+				}
+
+				return dataTransfer;
+			});
+
+			this._currentClipboardItem?.dataTransferPromise.cancel();
+			this._currentClipboardItem = { handle: handle, dataTransferPromise: promise };
+		}));
+
+		this._register(addDisposableListener(container, 'paste', async (e: ClipboardEvent) => {
+			const model = editor.getModel();
+			const selection = this._editor.getSelection();
+			if (!model || !selection || !e.clipboardData) {
+				return;
+			}
+
+			if (!this.arePasteActionsEnabled(model)) {
+				return;
+			}
+
+			const originalDocVersion = model.getVersionId();
+
+			const providers = this._languageFeaturesService.documentPasteEditProvider.ordered(model);
+			if (!providers.length) {
+				return;
+			}
+
+			const handle = e.clipboardData?.getData(vscodeClipboardMime);
+			if (typeof handle !== 'string') {
+				return;
+			}
+
+			e.preventDefault();
+			e.stopImmediatePropagation();
+
+			const dataTransfer = toVSDataTransfer(e.clipboardData);
+
+			if (handle && this._currentClipboardItem?.handle === handle) {
+				const toMergeDataTransfer = await this._currentClipboardItem.dataTransferPromise;
+				toMergeDataTransfer.forEach((value, key) => {
+					dataTransfer.set(key, value);
+				});
+			}
+
+			if (!dataTransfer.has(Mimes.uriList)) {
+				const resources = await this._clipboardService.readResources();
+				if (resources.length) {
+					const value = resources.join('\n');
+					dataTransfer.set(Mimes.uriList, {
+						value,
+						asString: async () => value,
+						asFile: () => undefined,
+					});
+				}
+			}
+
+			dataTransfer.delete(vscodeClipboardMime);
+
+			for (const provider of providers) {
+				const edit = await provider.provideDocumentPasteEdits(model, selection, dataTransfer, CancellationToken.None);
```

Comment:
fyi - there is `EditorStateCancellationTokenSource` which helps you check the document version and automatically cancel requests